lib/trip_dispatcher.rb

```diff
@@ -1,70 +1,160 @@
+      passenger = @passengers.find { |passenger| passenger.id == line[0].to_i }
 
-        users << User.new(input_data)
-      end
+      input_data = {}
+      input_data[:id] = line[0].to_i
+      input_data[:vin] = line[1]
+      input_data[:status] = line[2].to_sym
+      input_data[:name] = passenger.name
 
-      return users
+      driver = Driver.new(input_data)
+      drivers << driver
     end
+    return drivers
+  end
 
+  def load_users(filename)
+    users = []
 
-    def load_trips(filename)
-      trips = []
-      trip_data = CSV.open(filename, 'r', headers: true,
-                                          header_converters: :symbol)
+    CSV.read(filename, headers: true).each do |line|
+      input_data = {}
+      input_data[:id] = line[0].to_i
+      input_data[:name] = line[1]
+      input_data[:phone] = line[2]
+
+      users << User.new(input_data)
+    end
+    return users
+  end
+
+  def load_trips(filename)
+    trips = []
+    trip_data = CSV.open(filename, 'r', headers: true,
+      header_converters: :symbol)
 
       trip_data.each do |raw_trip|
         passenger = find_passenger(raw_trip[:passenger_id].to_i)
+        driver = find_driver(raw_trip[:driver_id].to_i)
 
         parsed_trip = {
           id: raw_trip[:id].to_i,
+          driver: driver,
           passenger: passenger,
-          start_time: raw_trip[:start_time],
-          end_time: raw_trip[:end_time],
+          start_time: Time.parse(raw_trip[:start_time]),
+          end_time: Time.parse(raw_trip[:end_time]),
           cost: raw_trip[:cost].to_f,
           rating: raw_trip[:rating].to_i
         }
 
         trip = Trip.new(parsed_trip)
         passenger.add_trip(trip)
+        driver.add_driven_trip(trip)
         trips << trip
       end
-
       return trips
     end
 
+    def find_driver(id)
+      check_id(id)
+      return @drivers.find { |driver| driver.id == id }
+    end
+
     def find_passenger(id)
       check_id(id)
       return @passengers.find { |passenger| passenger.id == id }
     end
 
+    def generate_available_drivers
+      available_drivers = @drivers.find_all do |driver|
+        driver.status == :AVAILABLE
+      end
+    end
+
+    def check_drivers_not_passenger?(user_id)
+      all_available_drivers = generate_available_drivers
+
+      available_drivers = all_available_drivers.find_all do |driver|
+        driver.id != find_passenger(user_id).id
+      end
+
+      raise ArgumentError.new("NO DRIVERS AVAILABLE") if available_drivers.empty?
+
+      return available_drivers
+    end
+
+    def find_no_trips_driver(user_id)
+      available_drivers = check_drivers_not_passenger?(user_id)
+      return available_drivers.find { |driver| driver.driven_trips.empty? }
+    end
+
+    def find_oldest_trip_driver(user_id)
+      available_drivers = check_drivers_not_passenger?(user_id)
+
+      find_most_recent_trips = available_drivers.map do |driver|
+        driver.driven_trips.max_by(&:end_time)
+      end
```

Comment:
The logic on line 117 might work well as a helper method in `Driver`, something like `Driver.most_recent_driven_trip`